python-sdk/nuscenes/nuscenes.py

```diff
@@ -40,113 +40,144 @@ def __init__(self,
+
+                elif name == 'sample_data':
+
+                    # Decorate sample_data with sensor information.
+                    for record in table:
+                        cs_record = self.get('calibrated_sensor', record['calibrated_sensor_token'])
+                        sensor_record = self.get('sensor', cs_record['sensor_token'])
+                        record['sensor_modality'] = sensor_record['modality']
+                        record['channel'] = sensor_record['channel']
+
+                    # Reverse-index samples with annotations.
+                    for record in table:
+                        if record['is_key_frame']:
+                            sample_record = self.get('sample', record['sample_token'])
+                            sample_record['data'][record['channel']] = record['token']
+
+                elif name == 'sample_annotation':
+
+                    # Decorate sample_annotation table with for category name.
+                    for record in table:
+                        inst = self.get('instance', record['instance_token'])
+                        record['category_name'] = self.get('category', inst['category_token'])['name']
+
+                    # Decorate sample with a list of annotation tokens.
+                    for ann_record in table:
+                        sample_record = self.get('sample', ann_record['sample_token'])
+                        sample_record['anns'].append(ann_record['token'])
+
+                # Populate token2ind.
+                self._token2ind[name] = dict()
+                for ind, member in enumerate(table):
+                    self._token2ind[name][member['token']] = ind
+
+            return self.tables[name]
+        else:
+            # Default behaviour: call __getattribute__
+            return self.__getattribute__(name)
+
+    def _load_table(self, table_name) -> dict:
         """ Loads a table. """
         with open(osp.join(self.table_root, '{}.json'.format(table_name))) as f:
             table = json.load(f)
         return table
 
-    def __make_reverse_index__(self, verbose: bool) -> None:
+    def _load_tables(self) -> None:
         """
-        De-normalizes database to create reverse indices for common cases.
-        :param verbose: Whether to print outputs.
+        Loads dataset tables and performs reverse indexing.
+        If lazy loading is used, certain tables will be loaded when they are accessed.
         """
 
-        start_time = time.time()
-        if verbose:
+        load_time = time.time()
+        if self.verbose:
+            print("======\nLoading NuScenes tables for version {} ...".format(self.version))
+
+        # Load tables that do not use lazy loading.
+        for table_name in self.table_names:
+            if table_name not in self.lazy_table_names:
+                table_content = self._load_table(table_name)
+                self.__setattr__(table_name, table_content)
+        self.tables = dict()
```

Comment:
I don't like that we are creating all these instance variables outside the constructor. Also look at `self._token2ind` below.